pkgdb/src/registry/floxpkgs.cc

```diff
@@ -0,0 +1,696 @@
+
+[[nodiscard]] static const std::filesystem::path
+getRulesFile()
+{
+  return nix::getEnv( "_PKGDB_NIXPKGS_RULES_JSON" ).value_or( RULES_JSON );
+}
+
+
+/* -------------------------------------------------------------------------- */
+
+/**
+ * @brief Create a temporary directory containing a `flake.nix` which wraps
+ *        @a nixpkgsRef, applying _rules_ from `rules.json`.
+ */
+std::filesystem::path
+createWrappedFlakeDir( const nix::FlakeRef & nixpkgsRef )
+{
+  /* Create a temporary directory to put the filled out template and rules file
+   * in. */
+  std::filesystem::path tmpDir = nix::createTempDir();
+  debugLog( "created temp dir for flake template: path=" + tmpDir.string() );
+  std::filesystem::path rulesFilePath = getRulesFile();
+  std::filesystem::copy( rulesFilePath, tmpDir / "rules.json" );
+
+  /* Fill out the template with the flake references and the rules file path. */
+  std::ofstream            flakeOut( tmpDir / "flake.nix" );
+  static const std::string flakeTemplate =
+#include "./floxpkgs/flake.nix.in.hh"
+    ;
+  std::istringstream flakeIn( flakeTemplate );
+  std::string        line;
+  while ( std::getline( flakeIn, line ) )
+    {
+      /* Inject URL */
+      if ( line.find( "@NIXPKGS_URL@" ) != std::string::npos )
+        {
+          line.replace( line.find( "@NIXPKGS_URL@" ),
+                        std::string( "@NIXPKGS_URL@" ).length(),
+                        nixpkgsRef.to_string() );
+        }
+
+      /* Inject rules */
+      if ( line.find( "@PKGDB_RULES_FILE@" ) != std::string::npos )
+        {
+          line.replace( line.find( "@PKGDB_RULES_FILE@" ),
+                        std::string( "@PKGDB_RULES_FILE@" ).length(),
+                        ( tmpDir / "rules.json" ).string() );
+        }
+      flakeOut << line << '\n';
+    }
+  flakeOut.close();
+  debugLog( "filled out flake template: flake_ref=" + nixpkgsRef.to_string()
+            + " rules_file_path=" + rulesFilePath.string() );
+
+  /* Lock the filled out template to avoid spurious re-locking and silence the
+   * "Added input ..." message. */
+  flox::NixState           nixState;
+  nix::ref<nix::EvalState> state = nixState.getState();
+  nix::flake::LockFlags    flags;
+  std::string              wrappedUrl = "path:" + tmpDir.string();
+  nix::FlakeRef            wrappedRef = nix::parseFlakeRef( wrappedUrl );
+  auto _locked = nix::flake::lockFlake( *state, wrappedRef, flags );
+  debugLog( "locked flake template" );
+
+  return tmpDir;
+}
+
+
+/* -------------------------------------------------------------------------- */
+
+/* Copied from the nix codebase and very slightly modified since these
+ * definitions aren't in header files.*/
+
+// A github, gitlab, or sourcehut host
+const static std::string hostRegexS = "[a-zA-Z0-9.-]*";
+std::regex               hostRegex( hostRegexS, std::regex::ECMAScript );
+
+std::optional<nix::fetchers::Input>
+GitArchiveInputScheme::inputFromURL( const nix::ParsedURL & url ) const
```

Comment:
I think this will be really confusing if you expect changes to the git*inputscheme to take effect globally, but they wont because they are not added to the global parser...

Reply:
I explicitly don't expect them to take effect globally and that's why I don't register them globally. These copy/pasted definitions are literally only here so that I can inherit from them. It would also be pointless to have two `Git*InputScheme`s registered because ours would be at the end of the list, and the one in Nix would always be picked up before ours.

Reply:
I'll add a comment to the header explaining that our definitions aren't registered.